src/box/lua/upgrade.lua

```diff
@@ -1523,6 +1536,390 @@ local function upgrade(options)
+            end
+        end
+    end
+end
+
+local function disable_core_field_foreign_key_constraints(issue_handler)
+    for _, row in box.space._space:pairs() do
+        for _, format in pairs(row.format) do
+            if format.foreign_key then
+                issue_handler(
+                    "Foreign key constraint is found in space '%s'," ..
+                    " field '%s'. It is supported starting from" ..
+                    " version 2.10.0.",
+                    row.name, format.name)
+            end
+        end
+    end
+end
+
+local function disable_core_tuple_foreign_key_constraints(issue_handler)
+    for _, row in box.space._space:pairs() do
+        if row.flags.foreign_key then
+            issue_handler(
+                "Foreign key constraint is found in space '%s'. " ..
+                "It is supported starting from version 2.10.0.",
+                row.name)
+        end
+    end
+end
+
+local function disable_core_field_constraints(issue_handler)
+    for _, row in box.space._space:pairs() do
+        for _, format in pairs(row.format) do
+            if format.constraint then
+                issue_handler(
+                    "Field constraint is found in space '%s', field '%s'. " ..
+                    "It is supported starting from version 2.10.0.",
+                    row.name, format.name)
+            end
+        end
+    end
+end
+
+local function disable_core_tuple_constraints(issue_handler)
+    for _, row in box.space._space:pairs() do
+        if row.flags.constraint then
+            issue_handler(
+                "Tuple constraint is found in space '%s'. " ..
+                "It is supported starting from version 2.10.0.",
+                row.name)
+        end
+    end
+end
+
+local function downgrade_from_2_10_0(issue_handler)
+    remove_deferred_deletes(issue_handler)
+    disable_background_space_upgrade(issue_handler)
+    disable_tuple_compression(issue_handler)
+    disable_core_field_foreign_key_constraints(issue_handler)
+    disable_core_tuple_foreign_key_constraints(issue_handler)
+    disable_core_field_constraints(issue_handler)
+    disable_core_tuple_constraints(issue_handler)
+end
+
+--------------------------------------------------------------------------------
+-- Tarantool 2.10.5
+--------------------------------------------------------------------------------
+
+local function drop_vspace_sequence_space(issue_handler)
+    if issue_handler.dry_run then
+        return
+    end
+    log.info("revoke grants for 'public' role for _vspace_sequence")
+    box.space._priv:delete{PUBLIC, 'space', box.schema.VSPACE_SEQUENCE_ID}
+    local indexes = box.space._index:select(box.schema.VSPACE_SEQUENCE_ID)
+    -- Otherwise we can't drop neither the primary index nor the space.
+    set_system_triggers(false)
+    for _, index in pairs(indexes) do
+        log.info("drop index %s on _vspace_sequence", index[3])
+        box.space._index:delete{index[1], index[2]}
```

Comment:
And here too - we must pcall these changes in order to enable system triggers back.

Reply:
Fixed.

```diff
     local indexes = box.space._index:select(box.schema.VSPACE_SEQUENCE_ID)
     -- Otherwise we can't drop neither the primary index nor the space.
     set_system_triggers(false)
-    for _, index in pairs(indexes) do
-        log.info("drop index %s on _vspace_sequence", index[3])
-        box.space._index:delete{index[1], index[2]}
+    local status, err = pcall(function()
+        for _, index in pairs(indexes) do
+            log.info("drop index %s on _vspace_sequence", index[3])
+            box.space._index:delete{index[1], index[2]}
+        end
+        log.info("drop view _vspace_sequence")
+        box.space._space:delete{box.schema.VSPACE_SEQUENCE_ID}
+    end)
+    if not status then
+        pcall(set_system_triggers, true)
+        error(err)
     end
-    log.info("drop view _vspace_sequence")
-    box.space._space:delete{box.schema.VSPACE_SEQUENCE_ID}
     set_system_triggers(true)
```